### ISSUE TITLE
feature: Fluentd in_tail plugin

### ISSUE THREAD

**Issue by jjsiv** [fluent/fluent-operator#632]
### Is your feature request related to a problem? Please describe.

At my org we would like to migrate some of the setups that we use Fluentd for to Fluentd with Fluent Operator. Currently it is not possible since we use in_tail plugin which is not available in Fluent Operator.

### Describe the solution you'd like

I'd like to see the in_tail plugin implemented in Fluent Operator Fluentd. Currently, there are a few things that should/must be in place before that can be done:

**1. Mounting volumes**
This I will raise a PR for soon. Since I wanted to also add volumeClaimTemplates, this required the update of controller-gen.

**2. Ability to set the value of '< match >' for filters and outputs**
This seems to be currently disabled
https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/filter/types.go#L19
So the idea would be to enable this. At the same time I suppose we could enable setting "id" as well.

**3. Ability to set tag for inputs**
The way processBody function operates at the moment is that it will skip adding "tag" to the config body. 
https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/params/model.go#L158
This works fine for non-inputs since you don't need/want to have both the tag and '< match >', but it also prevents from setting a custom tag for inputs. My idea would be to modify the conditional to only skip if the plugin is also not an input.

**4. Perhaps it would be also nice to have the ability to run Fluentd as a DaemonSet, similarly to how you can run FluentBit as a StatefulSet with Collector CRD**
This is something that I might look into in the future rather than now but I'm open to seeing your opinions on this

I'd like to see your thoughts/ideas on this and let me know if I'm missing something.

**Comment by benjaminhuo** [comment]
> **2. Ability to set the value of '< match >' for filters and outputs**
> This seems to be currently disabled
> https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/filter/types.go#L19
> So the idea would be to enable this. At the same time I suppose we could enable setting "id" as well.

For fluentd, Fluent Operator utilize `label router` plugin to achieve multi-tenant log distribution, you can find a generated fluentd config below, for each label id, the filter and output matches all logs for the namespaces the label id represents:
https://github.com/kubesphere-sigs/fluent-operator-walkthrough#how-to-check-the-configuration-and-data
```yaml
kubectl -n fluent get secrets fluentd-config -ojson | jq '.data."app.conf"' | awk -F '"' '{printf $2}' | base64 --decode 
---
<source>
  @type  forward
  bind  0.0.0.0
  port  24224
</source>
<match **>
  @id  main
  @type  label_router
  <route>
    @label  @48b7cb809bc2361ba336802a95eca0d4
    <match>
      namespaces  kube-system,default
    </match>
  </route>
</match>
<label @48b7cb809bc2361ba336802a95eca0d4>
  <filter **>
    @id  ClusterFluentdConfig-cluster-fluentd-config::cluster::clusterfilter::fluentd-filter-0
    @type  record_transformer
    enable_ruby  true
    <record>
      kubernetes_ns  ${record["kubernetes"]["namespace_name"]}
    </record>
  </filter>
  <match **>
    @id  ClusterFluentdConfig-cluster-fluentd-config::cluster::clusteroutput::fluentd-output-es-0
    @type  elasticsearch
    host  elasticsearch-master.elastic.svc
    logstash_format  true
    logstash_prefix  fluent-log
    port  9200
  </match>
  <match **>
    @id  ClusterFluentdConfig-cluster-fluentd-config::cluster::clusteroutput::fluentd-output-kafka-0
    @type  kafka2
    brokers  my-cluster-kafka-brokers.kafka.svc:9092
    topic_key  kubernetes_ns
    use_event_time  true
    <format>
      @type  json
    </format>
  </match>
</label>
```

**Comment by benjaminhuo** [comment]
> **4. Perhaps it would be also nice to have the ability to run Fluentd as a DaemonSet, similarly to how you can run FluentBit as a StatefulSet with Collector CRD**
> This is something that I might look into in the future rather than now but I'm open to seeing your opinions on this

Seems that adding the fluentd `in_tail` plugin makes sense only if fluentd is deployed as a daemonset .
If fluentd is deployed as a statefulset, you'll be not sure which nodes the fluentd pod will be scheduled to and so collecting the file of a random node doesn't seem correct or necessary.

In current design, fluentbit is used as a log collecting agent which is deployed as a daemonset, while fluentd is used as an aggregator which receives logs from the network or fluentbit and does some advanced namespace restrictions.

We're open to adding the fluentd daemonset mode if there're strong requirements and maintainers

**Comment by benjaminhuo** [comment]
> **3. Ability to set tag for inputs**
> The way processBody function operates at the moment is that it will skip adding "tag" to the config body.
> https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/params/model.go#L158
> This works fine for non-inputs since you don't need/want to have both the tag and '< match >', but it also prevents from setting a custom tag for inputs. My idea would be to modify the conditional to only skip if the plugin is also not an input.

@zhu733756 do you remember why treating the tag like this?

**Comment by wenchajun** [comment]
I think it's a good idea to add fluentd's `tail` plugin, I think there can be multiple deployment methods, deployment or statefulset or others.

**Comment by zhu733756** [comment]
> non-inputs since you don't need/want to have both the tag and '< match >', but it also prevents from setting a custom tag for inputs. My idea would be to modify the conditional to on

As @jjsiv  says.  Perhaps we need to modify the conditions.

> > **3. Ability to set tag for inputs**
> > The way processBody function operates at the moment is that it will skip adding "tag" to the config body.
> > https://github.com/fluent/fluent-operator/blob/master/apis/fluentd/v1alpha1/plugins/params/model.go#L158
> > This works fine for non-inputs since you don't need/want to have both the tag and '< match >', but it also prevents from setting a custom tag for inputs. My idea would be to modify the conditional to only skip if the plugin is also not an input.
> 
> @zhu733756 do you remember why treating tag like this?

**Comment by kaiohenricunha** [comment]
Adding customPlugin to a fluentd input type would be a nice feature to have too. 

I've been trying to integrate fluentd with prometheus and it seems that I'll hit a wall [here](https://docs.fluentd.org/monitoring-fluentd/monitoring-prometheus#step-3-expose-metrics-by-prometheus-input-plugin-via-http).

I don't know if it's possible to monitor fluentd at the current state of the fluent-operator. Any ideas?

**Comment by wenchajun** [comment]
> Adding customPlugin to a fluentd input type would be a nice feature to have too.
> 
> I've been trying to integrate fluentd with prometheus and it seems that I'll hit a wall [here](https://docs.fluentd.org/monitoring-fluentd/monitoring-prometheus#step-3-expose-metrics-by-prometheus-input-plugin-via-http).
> 
> I don't know if it's possible to monitor fluentd at the current state of the fluent-operator. Any ideas?

This is the input component and you can show how that mounted `secret` is written. As mentioned above, I think it would be better to add the input component functionality in fluentd as well.

**Comment by jjsiv** [comment]
Thanks for the input everyone, I've had the plugin implementation prepared for a while now, I will review & rebase it and try to get the PR in during the week. I will also open an issue for a discussion about running Fluentd as a daemonset.
